accepted/00000-rbac.md

```diff
@@ -0,0 +1,271 @@
+# RFC: Role-Based Access Control
+
+- Feature Name: Role-Based Access Control (RBAC)
+- Start Date: 2024-11-12
+- Author: @cbbayburt
+
+# Summary
+[summary]: #summary
+
+This RFC proposes implementing role-based access control (RBAC) in Uyuni to meet users' needs for fine-grained access control and configuration.
+The solution is extensible, integrates with existing features of the codebase, and is independent of the ongoing efforts to migrate from the Struts framework.
+
+
+# Motivation
+[motivation]: #motivation
+
+Currently, access control rules are defined by static role information and controlled individually within each action.
+By default, each new feature must define its own access rules; if no rules are explicitly defined, the feature is available to all authenticated users.
+
+This proposal moves access control rules to a centrally managed location.
+The code implementation is generic and relies on configurable access policy definitions, allowing new use cases and rule definitions to be addressed through configuration rather than code changes.
+The Uyuni Server deploys with sensible defaults and enables users to define custom access policies (user access groups) to meet their specific needs.
+
+The goal of access control is to answer the following question: "Can user U perform operation O on resource R?"
+
+Access control operates on three distinct levels, each with specific goals and constraints:
+
+1. **Authentication:** Identifies the user calling this action.
+2. **Authorization:** Determines if the user is allowed to call the endpoints for this action.
+3. **Accounting:** Assesses whether the user is permitted to operate on a specific entity.
```

Comment:
Maybe it is my bad English, but are you sure `Accounting` means this? Everywhere I looked, accounting is about measuring, logging etc.

From the sentence in the RFC I would assume accounting is some action which can block user access.

Reply:
This is a concept in user access control called Authentication, Authorization, and Accounting (AAA)
Here some links talking about it:
- https://community.fs.com/encyclopedia/aaa.html
- https://info.support.huawei.com/info-finder/encyclopedia/en/AAA.html
- https://www.strongdm.com/blog/aaa-security

Reply:
I'll try to rephrase this to be more clear

Reply:
> IIUC, this RBAC implementation will not help with something like proxy admin seeing only systems below that proxy, right?
> 
> Visibility will still be based only on organization level, correct?

This actually refers to "Accounting". No, at this point we won't control access at this level.

Reply:
Well not according to those links @rjmateus provided. Accounting is according to those links:

```
Accounting, the final process in the framework, is all about measuring what's happening within the network. As part of the protocol, it will collect and log data on user sessions, such as length of time, type of session, and resource usage. The value here is that it offers a clear audit trail for compliance and business purposes.
```

```
Accounting

Accounting records the used service type, start time, and data traffic to collect and record the network resource usage of the user for implementing time- or traffic-based accounting and network monitoring.
```

So to me this sound like Accounting it measuring and logging and no way it can decide user has access to this or that. It is simple passive step of logging.

Reply:
A now I'm trying to figure out from where I have read that statement :-|

Reply:
@aaannz thank you for pointing it out. Resource access should be one step of the `Authorization` process. However, This RFC is focusing only at the part `...perform operation O...`. This would be already a big step, since the role/access control is spread across all the code (from API definition, to database access factories), and this is something we must clean.
Next step can be find a better way to perform resource access check. What do you think?

Reply:
Sorry, it seems I mixed up all the terms. I rewrote the part so it's hopefully clearer now.